### PR TITLE
OLS-197: Reconcile logic in OLS Operator

### PR DESCRIPTION
**Pull request by vbelouso**
## Description

Reconcile the number of replicas in the CR. The deployment reflects the new replica count.
Reconcile deployment resources in the CR. The deployment reflects the new resource limits/requests.

## Type of change

- [ ] Refactor
- [x] New feature
- [ ] Bug fix
- [ ] CVE fix
- [ ] Optimization
- [ ] Documentation Update
- [ ] Configuration Update
- [ ] Bump-up dependent library

## Related Tickets & Documents

- Related Issue https://issues.redhat.com/browse/OLS-197
- Closes #

## Checklist before requesting a review

- [x] I have performed a self-review of my code.
- [x] PR has passed all pre-merge test jobs.
- [ ] If it is a core feature, I have added thorough tests.

## Testing
- Please provide detailed steps to perform tests related to this code change.
- How were the fix/results from this change verified? Please provide relevant screenshots or results.

### CODE REVIEW

File: internal/controller/utils.go [hunk 1]
```diff
@@ -16,6 +17,35 @@ func updateDeploymentAnnotations(deployment *appsv1.Deployment, annotations map[
...
+
+	if *a.Spec.Replicas != *b.Spec.Replicas {
+		a.Spec.Replicas = b.Spec.Replicas
+		return true
+	}
+
+	return false
+}
+
+func syncDeploymentResources(a, b *appsv1.Deployment) bool {
```
Comment (raptorsun) [inline internal/controller/utils.go:34]: `syncDeploymentResources` sounds like it is going to sync every container in the deployment, while here it updates only the first container. Though it is enough for now, if in future we add new containers in different order, the function cannot generalize for every deployment. 
We can use container name as key to merge 2 container lists.

Comment (raptorsun) [reply internal/controller/utils.go:34]: We may just need a function to update the resource settings in a container lists of a deployment: 
`func setDeploymentContainerResources(deployment *appsv1.Deployment, resources *corev1.ResourceRequirements, containerName string) error`

File: internal/controller/utils.go [hunk 2]
```diff
@@ -16,6 +17,35 @@ func updateDeploymentAnnotations(deployment *appsv1.Deployment, annotations map[
 	}
 }
 
+// syncDeploymentReplicas synchronizes the replicas field between two Deployment objects.
+func syncDeploymentReplicas(a, b *appsv1.Deployment) bool {
```
Comment (raptorsun) [inline internal/controller/utils.go:21]: Similar to `syncDeploymentResources`, we just need `func setDeploymentReplicas(deployment *appsv1.Deployment, replicas int) (bool, error)`

File: internal/controller/utils.go [hunk 3]
```diff
@@ -16,6 +17,35 @@ func updateDeploymentAnnotations(deployment *appsv1.Deployment, annotations map[
...
+	if a == nil || b == nil {
+		return false
+	}
+
+	existingResources := &a.Spec.Template.Spec.Containers[0].Resources
+	desiredResources := b.Spec.Template.Spec.Containers[0].Resources
+	if !apiequality.Semantic.DeepEqual(*existingResources, desiredResources) {
+		*existingResources = desiredResources
+		return true
+	}
```
Comment (xrajesh) [inline internal/controller/utils.go:44]: We can also move this DeepEqual checks to its own function. We may want to reuse for other deployment like Redis in the future.

File: internal/controller/utils.go [hunk 4]
```diff
@@ -16,6 +18,39 @@ func updateDeploymentAnnotations(deployment *appsv1.Deployment, annotations map[
...
+		*deployment.Spec.Replicas = replicas
+		return true
+	}
+
+	return false
+}
+
+// setDeploymentContainerResources sets the resource requirements for a specific container in a Deployment.
+func setDeploymentContainerResources(deployment *appsv1.Deployment, resources *corev1.ResourceRequirements, containerName string) bool {
+	containerIndex := getContainerIndex(deployment, containerName)
```
Comment (raptorsun) [inline internal/controller/utils.go:33]: we may log the incident of missing container (containerIndex == -1) and returns.

File: internal/controller/ols_app_server_deployment.go [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+	if existingDeployment.Annotations == nil ||
+		existingDeployment.Annotations[OLSConfigHashKey] != r.stateCache[OLSConfigHashStateCacheKey] {
+		updateDeploymentAnnotations(existingDeployment, map[string]string{
+			OLSConfigHashKey: r.stateCache[OLSConfigHashStateCacheKey],
+		})
+		changed = true
+	}
+
+	// Validate deployment replicas.
+	if setReplicas := setDeploymentReplicas(existingDeployment, *desiredDeployment.Spec.Replicas); setReplicas {
```
Comment (raptorsun) [inline internal/controller/ols_app_server_deployment.go:170]: assignement of `setReplicas` is not neccessary. it is equivalent to 
```go
if setDeploymentReplicas(existingDeployment, *desiredDeployment.Spec.Replicas) {
  changed = true
}
```
or even more simple
```go
changed = changed || setDeploymentReplicas(existingDeployment, *desiredDeployment.Spec.Replicas)
```

File: internal/controller/ols_app_server_deployment.go [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+	if existingDeployment.Annotations == nil ||
+		existingDeployment.Annotations[OLSConfigHashKey] != r.stateCache[OLSConfigHashStateCacheKey] {
+		updateDeploymentAnnotations(existingDeployment, map[string]string{
+			OLSConfigHashKey: r.stateCache[OLSConfigHashStateCacheKey],
+		})
+		changed = true
+	}
+
+	// Validate deployment replicas.
+	if setDeploymentReplicas(existingDeployment, *desiredDeployment.Spec.Replicas) {
```
Comment (vishnuchalla) [inline internal/controller/ols_app_server_deployment.go:170]: can we combine these 2 if into 1 using OR please? looks like a growing ladder IMHO.

**Comment by vishnuchalla** [comment]
Do we want to take unit tests for this change in a different PR?

**Comment by raptorsun** [comment]
/lgtm

I think we can put the tests in another PR :)

File: internal/controller/utils.go (hunk 3 above)
Comment (xrajesh) [reply internal/controller/utils.go:43]: just a nit - `func deploymentEquals(current, new *appsv1.Deployment) bool`  , we may be able to use for redis deployment check.

**Pull request merged**